Add April (101st) technical notes (#320)
tc39/tc39-notes: meetings/2024-04/april-08.md

SHN: All right, I sent out, which many of you may have seen, to all members of ECMA an email a little while ago. We have a vacant position in our executive committee, so we are looking for nominations. It is typical that we do this only once a year, but it is also very important to be active and have engagement from members. So we have made the decision through the executive committee that we would do a nomination in a non-year-end term. So if you have from ordinary members somebody you would like to nominate. or if somebody would like to self-nominate, I would be happy to receive it. Just email it to me. It’s a good opportunity to be very active in the governance and the activities of ECMA through the different technical committees that are also involved. That’s just a reminder.

> SHN: I also have a reminder on the ECMA approval. So we have already, thank you very much, the addition 15 preparation that already began, that was document was frozen very soon after the San Diego meeting. I’m just making a reminder with the dates here, and I don’t think I missed it, but if I have missed it, you may correct me, but for ECMA 402 I have not seen the freeze of that document. If it has been done, thank you. If it has not been done, then please remember that we are very close to the timeline where you want to manage the 60-day opt out and the 60-day review before the general assembly, which then takes place in June.

Reviewer: ```suggestion
SHN: I also have a reminder on the ECMA approval. So we have already, thank you very much, the Edition 15 preparation that already began, that document was frozen very soon after the San Diego meeting. I’m just making a reminder with the dates here, and I don’t think I missed it, but if I have missed it, you may correct me, but for ECMA 402 I have not seen the freeze of that document. If it has been done, thank you. If it has not been done, then please remember that we are very close to the timeline where you want to manage the 60-day opt out and the 60-day review before the general assembly, which then takes place in June.
```

Reviewer 2: correction correct.

SHN: We also have a new proposal that we will be discussing at the April ECMA ExeCom, which is TC55, pending the name, WinterCG. Again, if there’s interest from members of this committee on that, it’s excellent that we have a new technical topic to be working on. We already chartered the TC39 TG5 on the experience and programming languages. I hope that’s moving forward. I believe our colleagues from Bergen are on the call and will give an update, and this is very good, so I look forward to this continuing to move forward.

> SHN: Very happy to welcome some new members. I had noted them as potentials on my last meeting. We have received all the applications. So they all are provisionary members today. So replay IO, hero devs and sen tear, just little bit of a small document, but important one I’m waiting for the RF to be signed and provide back to my attention. And then there should be of course complete participation also from sentarc. We have new members and we have a big committee, I just wanted to bring to the attention that when an application comes to the ECMA international secretariat and we start the processes, the member company after very minor exchange of documents become provisionary members. They may participate with their delegates in any TC they wish. It is important to note that they may not vote until they become an official member, which typically takes place after the general assembly. Just want to highlight that right now so the members that are new that have joined, they are provisionary members, and they do not have voting rights. In addition, I want to mention that in all of these members that have joined, there have been invited experts from those organizations to become -- to be part of the TC and check it out which then eventually become members. Your invited expert designation will change to delegate designation once you become an official member, and that can take place in June time frame. Just to let you know that your designation from going from invited expert to a delegate does not in any way change the level of the expertise you bring into the team. Just the title of the designation is just slightly different. So that’s just any clarification.

Reviewer: ```suggestion
SHN: Very happy to welcome some new members. I had noted them as potentials on my last meeting. We have received all the applications. So they all are provisionary members today. So Replay.io, HeroDevs and Sentry, just little bit of a small document, but important one I’m waiting for the RF to be signed and provide back to my attention. And then there should be of course complete participation also from Sentry. We have new members and we have a big committee, I just wanted to bring to the attention that when an application comes to the ECMA international secretariat and we start the processes, the member company after very minor exchange of documents become provisionary members. They may participate with their delegates in any TC they wish. It is important to note that they may not vote until they become an official member, which typically takes place after the general assembly. Just want to highlight that right now so the members that are new that have joined, they are provisionary members, and they do not have voting rights. In addition, I want to mention that in all of these members that have joined, there have been invited experts from those organizations to become -- to be part of the TC and check it out which then eventually become members. Your invited expert designation will change to delegate designation once you become an official member, and that can take place in June time frame. Just to let you know that your designation from going from invited expert to a delegate does not in any way change the level of the expertise you bring into the team. Just the title of the designation is just slightly different. So that’s just any clarification.
```

> SHN: Again, as we have new members and we have new invited experts always coming on, and thank you, USA, you already very clearly mentioned that the chairs always menace at the beginning of every meeting that there a code of conduct committee. I just wanted to take the time, I’ve been through your website -- your GitHub page. I’ve just extracted some key words. I want to remind as the committee, as we work together, it’s very important that we work together in a very open and constructive manner on any of the channels that we are engaging for this meeting. Just as a reminder, I’ve put the up keywords we do tend to use this, I know we have lots of hot conversations, deep discussions, please always remember this, and continue to be as productive as we are.

Reviewer: ```suggestion
SHN: Again, as we have new members and we have new invited experts always coming on, and thank you, USA, you already very clearly mentioned that the chairs always mention at the beginning of every meeting that there is a Code of Conduct committee. I just wanted to take the time, I’ve been through your website -- your GitHub page. I’ve just extracted some key words. I want to remind as the committee, as we work together, it’s very important that we work together in a very open and constructive manner on any of the channels that we are engaging for this meeting. Just as a reminder, I’ve put the up keywords we do tend to use this, I know we have lots of hot conversations, deep discussions, please always remember this, and continue to be as productive as we are.
```

KG: Not much in the way of editorial changes to call out. The first one here, we’re calling out really only as an example of sort of a design principle. Which is that there is an abstract operation for finding the index of one string within another. Basically string.prototype.index of. And we tweaked it so that instead of returning negative one when the value wasn’t found, it returns a special value, and this is the approach that the editors intend to take for operations like this going forward, and if you’re writing a proposal and designing such an abstract operation yourself, we recommend this approach to you as well. It just generally makes it easier for readers and since the spec is not code, it’s not like there’s any overhead associated with returning a special value instead of returning negative 1. Not returning a number doesn’t actually affect anything. This isn’t C. So just something to keep in mind when designing things like this in the future.

> KG: Also, quite a few miscellaneous use cleanups and consistency fixes that we don’t feel are worth calling to the committee’s attention, but of curious you’re welcome ore view the commit history on GitHub. Not going to go through the list of upcoming work, because it hasn’t changed. We are still plugging away at a bunch of stuff. Most of these things we’re not actively working on. Some of them we are.

Reviewer: ```suggestion
KG: Also, quite a few miscellaneous use cleanups and consistency fixes that we don’t feel are worth calling to the committee’s attention, but if curious you’re welcome to view the commit history on GitHub. Not going to go through the list of upcoming work, because it hasn’t changed. We are still plugging away at a bunch of stuff. Most of these things we’re not actively working on. Some of them we are.
```

PFC: Okay, thanks. That’s good information as well.

PFC: All right. (Slide 9) I would like to request consensus on this pull request linked here that fixes the rounding bug that I described.

USA: All right. Let’s give it a minute. I would also reiterate that any sort of statements of explicit support are also welcome. All right. There doesn’t seem to be anything in the queue. So you have consensus.

PFC: Okay. Thanks. (Slide 10) I took the liberty of writing a proposed summary for the note, which I will show here and paste into the notes.

### Speaker's Summary of Key Points

- Consensus was reached on a normative change to fix a bug in rounding that occurred in rare cases having to do with DST.
- Over the next few weeks, we plan to dig into remaining concerns from TC39 delegates, particularly with the goal of reducing complexity.
- Follow the checklist in [#2628](https://github.com/tc39/proposal-temporal/issues/2628) for updates.

## Duplicate named capture groups for stage 4

Presenter: Kevin Gibbons (KG)

- [PR](https://github.com/tc39/ecma262/pull/2721)

> KG: Okay. So duplicate named capture groups. As a reminder, since it’s been a couple of years since I presented this, this is a feature that allows you to have the same capturing group name into two parts of a regular expression, with the constraint that they can’t both participate in the match, which is to say they have to be in different alternatives. So separated by a pipe. But otherwise, it works exactly like you would expect. You can use back references to the capturing group name. The.groups object of the result will contain the value from whichever one actually matched. In the case of repetition groups the last reputation group defines it, the same way it works for regular capturing groups.

Reviewer: ```suggestion
KG: Okay. So duplicate named capture groups. As a reminder, since it’s been a couple of years since I presented this, this is a feature that allows you to have the same capturing group name into two parts of a regular expression, with the constraint that they can’t both participate in the match, which is to say they have to be in different alternatives. So separated by a pipe. But otherwise, it works exactly like you would expect. You can use back references to the capturing group name. The.groups object of the result will contain the value from whichever one actually matched. In the case of repetition groups the last repetition group defines it, the same way it works for regular capturing groups.
```

> KG: The specification text is quite simple, although it’s not been reviewed by all of the other editors yet. That is, it’s not been reviewed as a pull request. But of course the specification text is approved as part of getting to Stage 3. It has been shipping in Safari for a while, and shipping in Chrome in 125, which isn’t stable yet. I believe it’s currently in the dev channel. Since this is only causing syntax to become legal which wasn’t previously legal, there isn’t much risk of web incapability. SpiderMonkey uses V8's RegExp engine as the underlying thing for their engine. They have to do a little bit of integration work to expose the new functionality. And I believe that work is underway. So it’s not yet shipping in Firefox.

Reviewer: ```suggestion
KG: The specification text is quite simple, although it’s not been reviewed by all of the other editors yet. That is, it’s not been reviewed as a pull request. But of course the specification text is approved as part of getting to Stage 3. It has been shipping in Safari for a while, and shipping in Chrome in 125, which isn’t stable yet. I believe it’s currently in the dev channel. Since this is only causing syntax to become legal which wasn’t previously legal, there isn’t much risk of web incompatibility. SpiderMonkey uses V8's RegExp engine as the underlying thing for their engine. They have to do a little bit of integration work to expose the new functionality. And I believe that work is underway. So it’s not yet shipping in Firefox.
```
